Machine ID: Add Protos for Bound Keypair Joining

timothyb89 opened the pull request:
This adds protos for the new Bound Keypair join method. Implementation will follow in additional PRs.

RFD: https://github.com/gravitational/teleport/pull/52546
Closes #53378

timothyb89 committed:
Machine ID: Add Protos for Bound Keypair Joining

This adds protos for the new Bound Keypair join method.
Implementation will follow in additional PRs.

RFD: https://github.com/gravitational/teleport/pull/52546
Closes #53378

timothyb89 committed:
Switch to SSH public key encoding for operator compatibility

The operator's crdgen doesn't support `bytes`, so switch to SSH
public key encoding.

timothyb89 committed:
Add generated terraform provider files

timothyb89 commented:
I'm reasonably confident this is complete, but want to avoid merging this until I have a decent start on #53373 so I can make breaking proto changes without creating linter trouble; will undraft when ready.

timothyb89 committed:
Add insecure flag

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bound-keypair-protos

timothyb89 committed:
Small proto tweaks

This adds some minor proto tweaks following some implementation work:
- Adds an explicit rotation ceremony phase to prompt the client to
  provide a new public key before issuing the 2nd challenge.
  Previously the design didn't include any way to inform the client
  that a rotation was requested.
- Converts byte public keys to string. They are now encoded in ssh
  `authorized_keys` format.

timothyb89 commented on api/proto/teleport/legacy/client/proto/joinservice.proto:
```diff
@@ -194,6 +194,116 @@ message RegisterUsingOracleMethodResponse {
+  // Any value set here will be ignored if a keypair is already bound.
+  string initial_join_secret = 2;
+
+  // If set, requests a rotation to the new public key. The joining challenge
+  // must first be completed using the previous key, and upon completion a new
+  // challenge will be issued for this key. Certificates will only be returned
+  // after the second challenge is complete. On initial join when using an
+  // initial join secret, this value should be set to the desired public key.
+  // This key is written in SSH public key format.
+  string new_public_key = 3;
```
Design question, my initial design was missing a way to inform the client that a rotation had actually been requested server-side. I've now added an explicit ceremony phase for this (`RegisterUsingBoundKeypairRotationRequest`) so the server will explicitly ask the client to provide a new public key, and then issue the 2nd challenge against it.

Anyway, with that mechanism in place, I think we could technically remove this field. Do we want to keep it and allow clients to initiate a key rotation on their own? I'm sort of leaning toward "no", now that we have the choice, but am curious for others' thoughts.

boxofrad replied:
I'd be in favor of removing it to simplify things, unless there's a compelling reason to allow clients to initiate the rotation?

timothyb89 replied:
I agree, and think I'll remove it. I'd rather have one flow now, and we can look into cleaner ways to implement client-driven rotation if that ever comes up. I don't think I like the complexity it adds to the challenge ceremony if we have to account for both options.

timothyb89 replied:
I've removed the field now, so rotation will now be exclusively driven by the server.

timothyb89 committed:
Update generated files

timothyb89 committed:
Fix docstring typo

timothyb89 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1892,6 +1897,145 @@ message ProvisionTokenSpecV2Oracle {
+  // RotateOnNextRenewal is a flag that indicates clients should perform a
+  // keypair rotation on the next refresh or rejoin attempt. This flag will
+  // automatically be set to `false` after a successful rotation.
+  bool RotateOnNextRenewal = 3 [(gogoproto.jsontag) = "rotate_on_next_renewal"];
```
Hmm, that's a fair point. Maybe we use some sort of a latch instead, something that will not require server-driven modification to the value? Like "rotate_after_timestamp", where a rotation is requested on the first join attempt after a specified timestamp?

That said, if server-driven resource updates will cause issues with IaC, I think there might be some broader problems. The `status` field has some fields important to Terraform, like the generated `initial_join_secret` value. My understanding is that Terraform should be able to handle this case without too much trouble, but maybe I wasn't quite correct about that?

timothyb89 replied:
Per discussion: we'll replace this with a timestamp field, e.g. `rotate_after: <timestamp>`. I don't think we see any critical issues with the Terraform provider around this (though the k8s operator might have trouble), but semantically we want to be clearly delegate updating `spec` to users, and `status` to Teleport.

 (there will eventually be a tctl ux for this)

timothyb89 replied:
This has been renamed and converted to a timestamp.

timothyb89 commented on api/proto/teleport/legacy/client/proto/joinservice.proto:
```diff
@@ -194,6 +194,116 @@ message RegisterUsingOracleMethodResponse {
+// RegisterUsingBoundKeypairChallengeResponse is the completed challenge
+// response sent by joining clients during the bound-keypair joining process.
+message RegisterUsingBoundKeypairChallengeResponse {
+  // A solution to a challenge from the server.
```
It's a signed JWT, so it's relatively self describing. But that's a fair call out - _that_ should at least be described. I'll add some detail.

timothyb89 replied:
I've added a rough description of the format of this field to the doc comment.

timothyb89 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1892,6 +1897,145 @@ message ProvisionTokenSpecV2Oracle {
+    // creation for the initial join attempt to succeed.
+    uint32 TotalJoins = 2 [(gogoproto.jsontag) = "total_joins"];
+
+    // MayJoinUntil is an optional timestamp before which all joining attempts
```
Not quite, it's different in a couple ways:
1. It only affects hard join attempts, i.e. when the internal identity has expired. Regular, "soft" joins act more like renewals, but still require this token to exist on the backend; this field doesn't apply to those.
2. Like with other aspects of this join method (e.g. the rejoin counter), it's intended to let things break but be recovered easily. If the regular `.metadata.expiry` field is used the token resource itself will be deleted automatically and will be difficult or annoying to recover; if this value elapses, it can be fixed by just adjusting the value again.

timothyb89 replied:
I've done a pass on the doc comments to hopefully improve clarity on this (among other things).

timothyb89 replied:
Per discussion, we'll remove this field for now since it's not really necessary. We can reevaluate later based on user feedback - it'll be easy to add later if it comes up.

timothyb89 replied:
I've removed this field.

timothyb89 committed:
Address review comments

bernardjkim commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1892,6 +1897,145 @@ message ProvisionTokenSpecV2Oracle {
+  // RemainingJoins is a count of the total number of additional joins that can
+  // be performed using this token. If `.spec.joining.total_joins` is
+  // incremented, this value will be incremented by the same value.
+  uint32 RemainingJoins = 4 [(gogoproto.jsontag) = "remaining_joins"];
```
Is this field needed? It feels like it is a bit redundant and adds a bit of complexity.

Does this count the number of remaining join attempts or successful joins? Would this essentially be the same value as `total_joins` - `join_count`?

Can `total_joins` be decremented? What happens if `total_joins` is decremented by more than the number of `remaining_joins` and then incremented again?

Might be confusing to the user if the `0` value can indicate that there are no more remaining joins or there are unlimited joins.

timothyb89 replied:
Huh, I think you're right. That removes an awkward calculated field and makes all the counters make sense regardless of whether or not `unlimited` is enabled.

For context, `remaining_joins` was originally the only field, and the zero value was just going to be a bit confusing if `unlimited` was set; `join_count` was only added later for informational purposes. But... I think you're right, using that field for everything is much simpler.

I'll make the change. Good catch!

timothyb89 replied:
> Can total_joins be decremented? What happens if total_joins is decremented by more than the number of remaining_joins and then incremented again?

It can be decremented, yes. Originally it was going to be an API error to decrement `total_joins` by more than the current value of `remaining_joins` (to maintain `remaining_joins >= 0` as an invariant).

With `remaining_joins` removed, though, I think it's actually okay to allow `total_joins < join_count` - that just means future join attempts will fail until `total_joins` is appropriately incremented.

timothyb89 replied:
I've gone ahead and removed `remaining_joins` in favor of `join_count`, like you suggested. I also did another pass on the doc comments to remove references to `remaining_joins` and improve clarity.

bernardjkim commented on api/proto/teleport/legacy/client/proto/joinservice.proto:
```diff
@@ -194,6 +194,119 @@ message RegisterUsingOracleMethodResponse {
+}
+
+// RegisterUsingBoundKeypairRotationResponse is the response sent from the
+// client containing
```
```suggestion
// client containing the public key.
```

timothyb89 committed:
Remove remaining_joins, add previous_instance_id, and update docs

This removes `remaining_joins` from the token status per a review
recommendation, in favor of determining whether or not a join is
allowed from the `join_count` field.

This also adds the referenced `previous_instance_id` field to the
bot instance proto.

Additionally, various doc comments have been updated for clarity.

boxofrad commented on api/proto/teleport/legacy/client/proto/joinservice.proto:
```diff
@@ -194,6 +194,119 @@ message RegisterUsingOracleMethodResponse {
+
+  // If set, attempts to bind a new keypair using an initial join secret.
+  // Any value set here will be ignored if a keypair is already bound.
+  string initial_join_secret = 2;
```
minor nit/no strong opinions: should these be `optional` fields?

timothyb89 replied:
Ah, good call. Made the change for this field, and removed `new_public_key` outright.

timothyb89 replied:
Bah. This file is built using protoc-gen-gogo which doesn't support field presence, so this just generates errors. I'm unfortunately reverting this change and we'll just stick to empty/nonempty for now :(

timothyb89 commented on api/proto/teleport/legacy/client/proto/joinservice.proto:
```diff
@@ -194,6 +194,119 @@ message RegisterUsingOracleMethodResponse {
+  // after the second challenge is complete. On initial join when using an
+  // initial join secret, this value should be set to the desired public key.
+  // This key is written in SSH public key format.
+  string new_public_key = 3;
```
Yeah, the other thread on this is related, I think. I'm inclined to just remove this now since I have since added a server-driven rotation request as part of the bidirectional stream flow. It now looks like:

1. Client sends initial request (w/ optional initial join secret + previous join state, if any)
2. Server sends challenge
3. Client replies with signed challenge
4. `rotate_on_next_renewal` is set, so the server now replies with a request for a new public key
5. The client generates a new public key and replies
6. The server sends a challenge for the new public key
7. The client sends the solution

tigrato replied:
This seems the correct way for me. Given we control the grpc server handler, the flow is clear who is asking what and under which conditions.

timothyb89 committed:
Remove `new_public_key` field in favor of server-driven rotation

timothyb89 committed:
Make initial_join_secret optional; fix field index after field removal

timothyb89 committed:
Revert optional field marker

protoc-gen-gogo doesn't support optional fields, so reverting this
change.

timothyb89 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1892,14 +1897,155 @@ message ProvisionTokenSpecV2Oracle {
+    // offer the same security assurances. This flag should be used with care,
+    // and RBAC rules should heavily restrict which resources this identity can
+    // access.
+    bool Insecure = 4 [(gogoproto.jsontag) = "insecure"];
```
Another discussion note: we decided to merge Unlimited and Insecure into an enum field, something conceptually like:

```proto
enum RecoveryMode {
  STANDARD,
  UNLIMITED,
  UNLIMITED_INSECURE
}

RecoveryMode mode = 4;
```

(names subject to further change for clarity; if I can come up with good, clear mode names, I'll opt for that instead)

I'll also need to gauge the YAML (and hcl/k8s/etc) representations here. I think in practice this will be a server-validated string value since enum encoding is ... bad.

timothyb89 replied:
I've tentatively gone with a string field with these possible values:
- `standard` - best security, recovery count is limited, and client state is verified
- `relaxed` - recovery count is not verified, client state is verified
- `insecure` - neither recovery count nor client state is verified

What do you think of these values and definitions?

strideynet replied:
Perfect!

timothyb89 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1892,14 +1897,155 @@ message ProvisionTokenSpecV2Oracle {
+    ];
+  }
+
+  message JoiningSpec {
```
Overall change per discussion, we've opted to standardize terminology:
* Instead of "joining and rejoining", "soft and hard joining", "consuming a join", etc - we'll instead call join attempts that increment `join_count` a "recovery".
* Onboarding is clear enough and stays as is. The first join technically counts as a "recovery" but that's an implementation detail I don't think end users will need to care much about.
  * (As noted above, `onboarding` itself may still be split to differentiate preregistered vs registered secrets, exact details WIP) 
* We could theoretically rename the concept of standard-case rejoining that does _not_ increment `join_count` of "joining" to "renewal", but all the remaining parameters here will be removed, so if we split the messages it'd be empty. We can sort what to actually call it if/when some need to add a param for that comes up, e.g. reintroducing `may_join_until`

I'll sort out the details today; there might be a new embedded message type to clearly split recovery-related parameters from

timothyb89 replied:
I've renamed this field to "recovery". `join_count` is now `recovery.limit`.

timothyb89 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1892,14 +1897,155 @@ message ProvisionTokenSpecV2Oracle {
+    // may be used. Attempts to register using an initial join secret after this
+    // timestamp will not be allowed. This may be modified after creation if
+    // necessary to allow the initial registration to take place.
+    google.protobuf.Timestamp MustJoinBefore = 3 [
+      (gogoproto.stdtime) = true,
+      (gogoproto.nullable) = true,
+      (gogoproto.jsontag) = "must_join_before,omitempty"
+    ];
+  }
```
From today's discussion: rename this to more explicitly associate it with `initial_join_secret`; possibly group them into their own sub-message. I'll play with the ergonomics of that today.

timothyb89 replied:
This field has been renamed to `must_register_before`

timothyb89 reviewed (commented):
@strideynet and I had a good discussion today and came up with a plan to further simplify all the options for this new token type, and to reduce overloaded and confusing terminology. I've left individual comments for all the changes for anyone interested.

This change will _mostly_ consist of field renames and moves, so hopefully nothing too surprising.

timothyb89 committed:
Field naming refactor

This refactors a number of bound keypair terms to simplify the token
and improve clarity.

We now consistently use these terms to refer to various concepts in
the join method:
- "onboarding" to refer to the first join
- "registration" to refer to the binding of a keypair to a token
- "preregistration" to refer to keys added manually with the token
- "registration secret" to refer to keys bound at the time of
  onboarding using a secret token
- "recovery" to refer to joining with no or an expired identity

Various fields were renamed to reflect these new terms and
definitions.

Additionally, the "unlimited" and "insecure" fields have been merged
into an enum-style string, with values "standard", "relaxed", and
"insecure".

timothyb89 committed:
Fix doc comment typo

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bound-keypair-protos
